### ISSUE TITLE
Testing on LBA model

### ISSUE THREAD

**Issue by DominiqueMakowski** [Julia-Tempering/Pigeons.jl#106]
I am testing Pigeons on an [LBA](https://github.com/itsdfish/SequentialSamplingModels.jl) Turing model but when recording the traces the process takes very long (it didn't go past 256 scans) and I was just wondering if that is expected or whether I am doing something wrong:

```julia
using Turing
using SequentialSamplingModels
using Random
using LinearAlgebra
using StatsPlots
using Random
using Pigeons

Random.seed!(45461)

# Generate some data with known parameters
dist = LBA(ν=[3.0, 2.0], A=0.8, k=0.2, τ=0.3)
data = rand(dist, 100)

# Specify LBA model
@model function model_lba(data; min_rt=minimum(data.rt))
    # Priors
    ν ~ MvNormal(zeros(2), I * 2)
    A ~ truncated(Normal(0.8, 0.4), 0.0, Inf)
    k ~ truncated(Normal(0.2, 0.2), 0.0, Inf)
    τ ~ Uniform(0.0, min_rt)

    # Likelihood
    data ~ LBA(; ν, A, k, τ)
end

# MCMC sampling takes ~20 s
# chain = sample(model_lba(data), NUTS(1000, 0.85), MCMCThreads(), 1000, 4)

# With Pigeons
pt = pigeons(target=TuringLogPotential(model_lba(data)), record=[traces])
samples = Chains(sample_array(pt), variable_names(pt))
```

**Comment by alexandrebouchard** [comment]
Here is my guess at what happens: in contrast to HMC, PT will sample from the prior at each iteration (at the reference chain). Since the model seems to use hitting time type processes, is it possible that for some parameter values plausible under prior but not posterior, the process never hits? For example there is a potentially infinity loop there: https://github.com/itsdfish/SequentialSamplingModels.jl/blob/28f5b626eadb72abafbc28ac64bc7dfc34ebb6b4/src/LBA.jl#L76
Let me know if that's a reasonable hypothesis and we can help brainstorm potential solutions (eg with variational PT).

**Comment by DominiqueMakowski** [comment]
Interesting, thanks for digging into this! As I'm out of my depth here, tagging @itsdfish for input 😁

**Comment by itsdfish** [comment]
Oh, and I just realized that the function Alexander pointed to is not part of the likelihood function. So that definitely rules that out. Sorry I didn't realize that sooner.

**Comment by miguelbiron** [comment]
It seems that the issue is related to numerical instability in the calculation of the loglikelihood. The default explorer in Pigeons is [SliceSampler](https://doi.org/10.1214/aos/1056562461) with an interval-doubling strategy. By default, it does up to a maximum of 20 doublings to find a candidate point in the current potential slice. This is an aggressive but desirable strategy, as it improves mixing. However, it seems to fail in this model. You can make it work by reducing the maximum number of doublings to 3, via
```julia
pt = pigeons(
    target=TuringLogPotential(model_lba(data)),
    explorer = SliceSampler(p=3),
    record=[traces]
)
```

I haven't been able to pinpoint the exact problem, but my suspicion starts with the implementation of [logpdf as log(pdf(x))](https://github.com/itsdfish/SequentialSamplingModels.jl/blob/28f5b626eadb72abafbc28ac64bc7dfc34ebb6b4/src/LBA.jl#L95). This usually induces a massive loss of numerical precision, compared to working directly on log scale. It is likely that the aggressive proposals of the default explorer are getting somehow accepted simply because of loss of precision. And so the sampler drifts towards infinity. I need to run more tests to confirm this suspicion though.

**Comment by miguelbiron** [comment]
Ok I think I found the real underlying issue. The model has a degeneracy with ~ infinite likelihood in this corner case
```julia
julia> ν = [-2.749666828178191, -2.975375043972598];

julia> A, k, τ = (2.4522048137701057e-308, 1.9900766988146945, 4.1409819816881695e-8);

julia> lba=LBA(; ν, A, k, τ)
LBA
┌───────────┬──────────────────────────────────────────┐
│ Parameter │ Value                                    │
├───────────┼──────────────────────────────────────────┤
│ ν         │ [-2.749666828178191, -2.975375043972598] │
│ A         │  0.00                                    │
│ k         │  1.99                                    │
│ τ         │  0.00                                    │
│ σ         │ [1.0, 1.0]                               │
└───────────┴──────────────────────────────────────────┘


julia> loglikelihood(lba, data)
53678.092133208265
```
Evidently, the Slice Sampler with the doubling strategy is the only one finding this infinite mode. Stepping out and HMC are missing it. 

The problem seems to be that exploration around this mode takes forever because you can go as close to neg infty in log(A) and log(tau) -- in transformed space -- as you want, so Slice sampler executes all the maximum doublings all the time.

**Comment by itsdfish** [comment]
Thank you for looking into the issue. If I understand correctly, the issue is a limitation of the Slice sampler. Is that the case?

**Comment by miguelbiron** [comment]
No, the issue is due to the mode described above, which seems to be caused entirely due to numerical instability in the calculation of the loglikelihood. In other words, at least from a modeller's perspective, it is suspicious that this weird LBA
```
┌───────────┬──────────────────────────────────────────┐
│ Parameter │ Value                                    │
├───────────┼──────────────────────────────────────────┤
│ ν         │ [-2.749666828178191, -2.975375043972598] │
│ A         │  0.00                                    │
│ k         │  1.99                                    │
│ τ         │  0.00                                    │
│ σ         │ [1.0, 1.0]                               │
└───────────┴──────────────────────────────────────────┘
```
fits the data infinitely better than the data-generating one. Indeed, the latter has posterior logdensity of about 38.8, whereas the former has posterior logdensity of 53627 --  >1000 times higher (in log scale!). Of course, its probably not the case that the weird model actually fits better. It's much more likely that this is due to numerical errors happening within the loglikelihood function for values of A and tau close to 0.

I suggest looking into why the loglikelihood function returns such high values for these combinations of parameters. Ultimately, Pigeons worked as expected in the face of multi-modality -- even in this case where the mode is almost surely due to an error.

**Comment by itsdfish** [comment]
Thanks for clarifying. Indeed. There are some strange things happening. First, as you noted, the log likelihood is very  large given how far the parameters are from the data generating parameters. Second, the parameters above strange relative to what is expected from the model. For example, the nu parameters are extremely negative, and A is typically greater than zero.

I noticed a similar problem when I constrained nu to be positive:

`ν ~ filldist(truncated(Normal(1, 1),0, Inf), 2)`

Out of curiosity, do you see a similar problem with that constraint? If so, that could suggest that the bounds of nu are not respected (perhaps a problem Bijectors), or there is a point in the more reasonable part of the parameter space that is producing anomalous results also.

**Comment by itsdfish** [comment]
I think I was able to locate the source of the error. Indeed, there are some numerical issues, but they may not have a complete solution because there are negative terms in the pdf. As a simple example,  log(-v * a + b * c) might provide a negative number close to zero due to numerical under/overflow. If there were not negative terms, we could use `logsumexp([log(v) + log(a),log(b)+log(c)])`, but I'm not sure if there is a workaround. 

I would like to test some solutions, but I am blocked by a dependency  conflict between Pigeons and SequentialSamplingModels. SequentialSamplingModels is not compatible with DynamicPPL < 0.22.0. Would it be possible for you to support 0.22.0 and/or 0.23.0?

**Comment by miguelbiron** [comment]
I just created PR #107 to batch update multiple compat entries. Should be merged soon after tests pass.

**Comment by itsdfish** [comment]
Thank you for upgrading. The results I am obtaining seem reasonable to me. In particular, the posterior distribution statistics are similar to those obtained from NUTS.

```
┌ Warning: More than one threads are available, but explore!() loop is not parallelized as pt.inputs.multithreaded == false
└ @ Pigeons ~/.julia/packages/Pigeons/BJYuc/src/pt/checks.jl:12
────────────────────────────────────────────────────────────────────────────
  #scans       Λ      log(Z₁/Z₀)   min(α)     mean(α)    min(αₑ)   mean(αₑ) 
────────── ────────── ────────── ────────── ────────── ────────── ──────────
        2       3.79       -111  7.63e-121      0.579          1          1 
        4       2.75       32.7   0.000353      0.694          1          1 
        8       2.72       35.3      0.275      0.698          1          1 
       16       3.69         38      0.364       0.59          1          1 
       32       3.57       35.7      0.414      0.603          1          1 
       64       3.58         36      0.438      0.603          1          1 
      128       3.88       35.7      0.526      0.569          1          1 
      256        3.7       35.4      0.496      0.589          1          1 
      512       3.79       35.5      0.524      0.579          1          1 
 1.02e+03       3.73       35.4      0.552      0.585          1          1 
────────────────────────────────────────────────────────────────────────────
Chains MCMC chain (1024×5×1 Array{Float64, 3}):

Iterations        = 1:1:1024
Number of chains  = 1
Samples per chain = 1024
parameters        = ν1, ν2, A, k, τ

Summary Statistics
  parameters      mean       std      mcse   ess_bulk   ess_tail      rhat   ess_per_sec 
      Symbol   Float64   Float64   Float64    Float64    Float64   Float64       Missing 
        ν[1]    2.8128    0.4162    0.0214   371.5544   593.1407    1.0009       missing
        ν[2]    1.6257    0.3542    0.0188   359.1784   600.5590    1.0021       missing
           A    0.7217    0.1756    0.0090   377.3798   505.1072    1.0076       missing
           k    0.2358    0.1142    0.0079   202.5119   304.5715    1.0144       missing
           τ    0.2818    0.0274    0.0019   215.7482   309.8607    1.0170       missing

```
One thing I noticed was that  `variable_names` does not destructure vectors. For example:

```
variable_names(pt)
4-element Vector{String}:
 "ν"
 "A"
 "k"
 "τ"
```
Would you like me to open a separate issue?

Finally, I am interested in obtaining the marginal log likelihood from the output above. I understand Z1/Z0 is a ratio. Can you recommend a way to obtain Z0 so that Z1 can be obtained? Thanks!

**Comment by miguelbiron** [comment]
I'm glad it's working. Did you fix the numerical issues or put a different prior?

Re: the variables names, yes please open a different issue.

Finally, re the normalizing constant, for turing models the prior should be normalized, so Z0=1. Have a look at [the docs age on normalizing constants](https://julia-tempering.github.io/Pigeons.jl/dev/output-normalization/#output-normalization) to see how to extract it.

**Comment by itsdfish** [comment]
I believe I fixed the numerical issue, or at least most cases in which it could arise. I'll open a new issue. 

Thanks again for your help in diagnosing the problem. Feel free to close this issue.

**Comment by miguelbiron** [comment]
Awesome!
